src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightDelegateCreator.cs

```diff
@@ -27,6 +27,7 @@ internal LightDelegateCreator(Interpreter interpreter, LambdaExpression lambda)
 
         public Delegate CreateDelegate() => CreateDelegate(closure: null);
 
+        [PreserveDependency("System.Linq.Expressions.Interpreter.LightLambda")]
```

Comment:
I don't see how this should help the method calls `LightLambda` hence has a dependency on the type already

Reply:
We need to preserve everything on the `LightLambda` type in this case. It has about 60 methods that are accessed via reflection. I guess it would be possible to list them all here, but that would be pretty unwieldy.

I _thought_ this change would do that, but it turns out my testing was incorrect. As you mentioned, this change won't preserve anything additional. In fact, it looks like `PreserveDependency` is only implemented for fields and members (although maybe we need to pull in new code from the upstream linker).

How should `PreserveDependency` for a type? Should it preserve merely the type, or should it act like `preserve="all"` in the link.xml file? In this case, I need the latter behavior.

Reply:
That's not how this attribute is intended to work. There is no weak reference from `LightDelegateCreator` to `LightLambda` hence the attribute is not needed here at all.

What you should do instead is to go to `LightLambda` and apply the attribute where needed. For example `MakeRunDelegateCtor` could have

```c#
#if NO_FEATURE_STATIC_DELEGATE
[PreserveDependency("RunVoidRef2")]
[PreserveDependency("RunVoid0")]
[PreserveDependency("Run0")]
[PreserveDependency("Run1")]
...
#endif
```

Now you'd need to list all the dependencies which we could add support for wildcard in linker to make it less painful but preserving whole type is exactly opposite of what this attribute was designed for.

Reply:
I've force pushed a new new changeset that add the attributes for each method to the `MakeRunDelegateCtor` method as you suggested.